docs/proposals/20250513-chained-and-efficient-upgrades-for-clusters-with-managed-topologies.md

```diff
@@ -0,0 +1,408 @@
+
+More specifically:
+- `BeforeClusterUpgrade` request will include the upgrade plan.
+- A new `BeforeControlPlaneUpgrade` hook will be added and called before each "upgrade control plane" step;
+  request and response payload will be similar to corresponding messages for `BeforeClusterUpgrade`
+- `AfterControlPlaneUpgrade` hook will now be called after each "upgrade control plane" step.
+- A new `BeforeWorkersUpgrade` hook will be added and called before each "upgrade workers" step;
+  request and response payload will be similar to corresponding messages for `BeforeControlPlaneUpgrade`
+- A new `AfterWorkersUpgrade` hook will be added and called after each "upgrade workers" step;
+  request and response payload will be similar to corresponding messages for `AfterControlPlaneUpgrade`, but the
+  hook will be considered blocking only for the intermediate steps of the upgrade (not blocking for the final step).
+- `AfterClusterUpgrade` will remain as of today, but the system will ensure that a new upgrade
+  can't start until `AfterClusterUpgrade` is completed.
+
+### Security Model
+
+This proposal does not add additional security concern to Cluster API.
+
+### Risks and Mitigations
+
+- Upgrading a Cluster by multiple Kubernetes minor versions in a short timeframe might increase risks to face issues during the upgrade.
+
+This proposal aims to help users those risks by automating the chained upgrade workflow so users can catch up with 
+Kubernetes versions easily, quickly, and with an upgrade plan validated by the system.
+
+Also, worth to notice that each machine rollout in Cluster API ultimately is an operation that is exercising 
+the same machinery that will be used during upgrades. 
+
+That means that by doing any rollout, e.g. due to an automatic machine remediation, you get a proxy signal about the
+fact that the system can successfully perform an upgrade, or you get the chance to detect and fix issues in the system
+before a full upgrade is performed.
+
+Conversely, risk increase for users not performing any form of rollouts for long periods.
+
+- Upgrading a Cluster by multiple Kubernetes minor versions might compromise workloads.
+
+Each Kubernetes minor upgrade might include changes impacting workloads running in the Cluster, and going through
+multiple minors increases the chance that workloads running in a cluster are impacted.
+
+As a mitigation, this proposal is improving lifecycle hooks and giving full visibility in the upgrade plan, thus
+allowing users to implement additional pre-upgrade checks.
+
+## Alternatives
+
+An alternative to option leveraging a new CRs to define the list of Kubernetes version to be used for upgrade plans
+was considered.
+
+However, the option was discarded because it seems more consistent having the list of 
+Kubernetes version to be used for upgrade plans in ClusterClasses, alongside all the other info defining 
+how a managed topology should behave.  
```

Comment:
```suggestion
Kubernetes version to be used for upgrade plans in ClusterClasses or calculated on-demand, alongside all the other info defining 
how a managed topology should behave.  
```

Reply:
can't we avoid this detail? (the motivation in this sentence mostly apply to the list of K8s versions, it is less relevant for the runtime extension configuration)